polybar: add module

uvNikita opened the pull request:
This module can be used only to configure polybar, but it doesn't start the bars.
The problem with launch is that it might require setting some environment variables depending on the user's configuration.
For example, this is the script I use to run my bar on all screens:

```
#!/usr/bin/env bash

killall -q polybar

while pgrep -u $UID -x polybar >/dev/null; do sleep 1; done

PRIMARY_MONITOR=`xrandr | grep "primary" | cut -d " " -f1`
MONITOR=$PRIMARY_MONITOR TRAY_POSITION=right polybar bar &

EXTRA_MONITORS=`xrandr | grep " connected" | grep -v $PRIMARY_MONITOR | cut -d " " -f1`
for MONITOR in $EXTRA_MONITORS; do
    MONITOR=$MONITOR polybar bar &
done

echo "Bars launched..."
```

uvNikita commented on modules/programs/polybar.nix:
```diff
@@ -0,0 +1,98 @@
+          generators.toINI {} cfg.config;
+      })
+
+      (mkIf (!lib.isAttrs cfg.config) {
```
Apparently, there is no such thing as `lib.isPath`, couldn't find a better way to check this condition.

rycee commented on modules/programs/polybar.nix:
```diff
@@ -0,0 +1,98 @@
+        defaultText = "pkgs.polybar";
+        description = "Polybar package to install.";
+        example =  literalExample ''
+          (pkgs.polybar.override {
```
I think the parentheses here are not necessary.

rycee commented on modules/programs/polybar.nix:
```diff
@@ -0,0 +1,98 @@
+        '';
+        default = {};
+        example = literalExample ''
+          config = {
```
I think the `config = ` part should be removed here. Typically the example is just what you would put to the right of the equals.

rycee commented on modules/programs/polybar.nix:
```diff
@@ -0,0 +1,98 @@
+      })
+
+      {
+        home.file.".gitconfig".text = cfg.extraConfig;
```
Huh? I'm a bit surprised to see `.gitconfig` here :-)

I'm guessing the extra config should be appended `.config/polybar/config"` but that seems tricky to me in the case where a user configures `programs.polybar.config` to be a path. You'd have to do something like

```
home.file.".config/polybar/config".text = readFile cfg.config + "\n" + cfg.extraConfig;
```

Looking at the Polybar wiki I saw the "[include-file](https://github.com/jaagr/polybar/wiki/Configuration#file-inclusion)" option. Perhaps an alternative is to set the type of `config` to

```
types.coercedTo
  types.path
  (p: { "section/base" = { include-file = p; }; })
  (types.attrsOf types.attrs);
```

Of course, I haven't tested this idea in the slightest ;-) But if it works then you shouldn't have problems with needing to identify whether `config` is an attribute set or a path and I guess the module's `config` section could be just

```
  config = mkIf cfg.enable {
    home.packages = [ cfg.package ];
    home.file.".config/polybar/config".text =
      generators.toINI {} cfg.config + "\n" + cfg.extraConfig;
  };
```

rycee commented:
I wonder if it would be sufficient to have a `script` option of type `str` that by default contains a very basic start-up command, maybe simply `script = "${cfg.package}/bin/polybar"`? Then create a `systemd.user.services.polybar` roughly like

```
systemd.user.services.polybar = {
  Unit = {
    Description = "Polybar desktop bar";
    After = [ "graphical-session-pre.target" ];
    PartOf = [ "graphical-session.target" ];
  };

  Service = {
    Type = "forking";
    Environment = "PATH=${cfg.package}/bin";
    ExecStart = pkgs.writeShellScriptBin "polybar-start" cfg.script;
  };

  Install = {
    WantedBy = [ "graphical-session.target" ];
  };
};
```
It should be then possible to override `script` for more advanced setups (like yours), with any luck systemd will be able to track all running polybars and stop them when the X session terminates :-)

Just a wild idea, I'm not certain it will work in practice…

uvNikita commented:
Thank you for the suggestions, it made the module much cleaner and better!

- Added news entry
- Fixed examples
- Implemented and tested your suggestion with `include-file`
- Created systemd service as you suggested

`pkgs.writeShellScriptBin` is not working properly right now (see https://github.com/NixOS/nixpkgs/pull/30153), so used `pkgs.writeScriptBin` for now.

Tested service start on this config:
```
    startScript = ''
      #!${pkgs.bash}/bin/bash
      PATH=$PATH:${pkgs.i3-gaps}/bin:${pkgs.xorg.xrandr}/bin:${pkgs.gnugrep}/bin:${pkgs.coreutils}/bin

      PRIMARY_MONITOR=`xrandr | grep "primary" | cut -d " " -f1`
      MONITOR=$PRIMARY_MONITOR TRAY_POSITION=right polybar bar &

      EXTRA_MONITORS=`xrandr | grep " connected" | grep -v $PRIMARY_MONITOR | cut -d " " -f1`
      for MONITOR in $EXTRA_MONITORS; do
          MONITOR=$MONITOR polybar bar &
      done
    '';
```
Works as it should, even with multiple screens. It would be also neat if home-manager would suggest restarting the service on polybar config file changes.

uvNikita commented:
Unfortunately, we can't set a default value for `startScript` parameter since it requires the name of the bar defined in the configuration file.

rycee commented on modules/services/polybar.nix:
```diff
@@ -0,0 +1,116 @@
+          It can be assumed that 'polybar' executable is in the PATH.
+        '';
+        example = ''
+          #!${pkgs.bash}/bin/bash
```
Maybe could move this shebang into the `writeScriptBin` below? Or maybe somebody would prefer a non-bash start script?

rycee commented on modules/services/polybar.nix:
```diff
@@ -0,0 +1,116 @@
+        '';
+        default = {};
+        example = literalExample ''
+          "bar/top" = {
```
Should wrap this whole thing in `{` and `}`. Basically a user should be able to copy the whole example, paste at `×` in `services.polybar.config = ×`, and have the evaluation succeed.

rycee commented on modules/services/polybar.nix:
```diff
@@ -0,0 +1,116 @@
+      };
+
+      config = mkOption {
+        # type = types.either (types.attrsOf types.attrs) types.path;
```
I think you can remove this line.

rycee commented on modules/services/polybar.nix:
```diff
@@ -0,0 +1,116 @@
+        description = ''
+          This script will be used to start the polybars.
+          Set all necessary environment variables here and start all bars.
+          It can be assumed that 'polybar' executable is in the PATH.
```
Can replace `'polybar'` by `<command>polybar</command>` and `PATH` by `<envar>PATH</envar>`.

rycee commented:
Nice work! I added a few final minor comments. Generally I think this is good to merge 👍

> It would be also neat if home-manager would suggest restarting the service on polybar config file changes.

Yes indeed! Unfortunately this is a bit tricky but I dearly hope that the future NixUP work will make it possible for us to use the NixOS systemd module and its nice `restartTriggers` option. I suppose some custom solution could be made here, basically you could to do something like what is done for the xmonad module but instead of doing `xmonad --restart` do `systemctl --user restart polybar`…

uvNikita commented:
Again, thank you for the comments!
Fixed everything you mentioned and implemented polybar service restart in the similar way as it's done for xmonad. Let me know if you think it's too hacky.

As for the script, I'm now adding shebang automatically since `writeShellScriptBin` is now fixed in master branch and once it will get to 17.09 (not sure when it will happen though), we can easily replace `writeScriptBin` with `writeShellScriptBin` without users noticing.

rycee commented on modules/services/polybar.nix:
```diff
@@ -0,0 +1,139 @@
+      fi
+    '';
+
+    home.activation.applyPolybar = dagEntryAfter [ "linkGeneration" ] ''
```
The
```
home.activation.checkPolybar = dagEntryBefore [ "linkGeneration" ]
```
above is OK but
```
home.activation.applyPolybar = dagEntryAfter [ "linkGeneration" ]
```
here should be
```
home.activation.applyPolybar = dagEntryAfter [ "reloadSystemD" ]
```
since the restart should be done after we're certain that systemd knows about the service file :-)

rycee commented:
Looks fine to me! I added one final comment but then feel free to merge once you are ready. Thanks!

I might look at some more general restart trigger mechanism but i think this is OK for now.

uvNikita commented:
Changed to `reloadSystemD` and rebased into master: https://github.com/rycee/home-manager/commit/469caa1a1405b22a849f60280ca4bdaec2f3289b
